packages/langium/src/generator/generator-tracing.ts

```diff
@@ -0,0 +1,196 @@
+            );
+            return region && copyDocumentSegment(region, getDocumentURI(astNode));
+
+        } else if (astNode.$cstNode) {
+            const region = getSingleOrCompoundRegion(
+                findNodesForProperty(astNode.$cstNode, property)
+            );
+            return region && copyDocumentSegment(region, getDocumentURI(astNode));
+
+        } else {
+            return undefined;
+        }
+    }
+}
+
+function getDocumentURI(astNode: AstNodeWithTextRegion): string | undefined {
+    if (astNode.$cstNode) {
+        return getDocument(astNode)?.uri?.toString();
+
+    } else if (astNode.$textRegion) {
+        return astNode.$textRegion.documentURI
+            || new TreeStreamImpl(astNode, n => n.$container ? [ n.$container ] : []).find( n => n.$textRegion?.documentURI )?.$textRegion?.documentURI;
+
+    } else {
+        return undefined;
+    }
+}
+
+function copyDocumentSegment(region: SourceRegionPartial, fileURI?: string): DocumentSegmentWithFileURI {
+    const result = <DocumentSegmentWithFileURI>{
+        offset: region.offset,
+        end:    region.end ?? region.offset + region.length!,
+        length: region.length ?? region.end! - region.offset,
+    };
+
+    if (region.range) {
+        result.range = region.range;
+    }
+
+    fileURI ??= region.fileURI;
+    if (fileURI) {
+        result.fileURI = fileURI;
+    }
+
+    return result;
+}
+
+function mergeDocumentSegment(prev: SourceRegionPartial | undefined, curr: SourceRegionPartial): DocumentSegmentWithFileURI {
+    if (!prev) {
+        return curr && copyDocumentSegment(curr);
+    } else if (!curr) {
+        return prev && copyDocumentSegment(prev);
+    }
+
+    const prevEnd = prev.end ?? prev.offset + prev.length!;
+    const currEnd = curr.end ?? curr.offset + curr.length!;
+    const offset = Math.min(prev.offset, curr.offset);
+    const end = Math.max(prevEnd, currEnd);
+    const length = end - offset;
+
+    const result = <DocumentSegmentWithFileURI>{
+        offset, end, length,
+    };
+
+    if (prev.range && curr.range) {
+        result.range = <Range>{
+            start: curr.range.start.line < prev.range.start.line
+                    || curr.range.start.line === prev.range.start.line && curr.range.start.character < prev.range.start.character
+                ? curr.range.start : prev.range.start,
+            end: curr.range.end.line > prev.range.end.line
+                    || curr.range.end.line === prev.range.end.line && curr.range.end.character > prev.range.end.character
+                ? curr.range.end : prev.range.end
+        };
+    }
+
+    if (prev.fileURI || curr.fileURI) {
+        const prevURI = prev.fileURI;
+        const currURI = curr.fileURI;
+        const fileURI = prevURI && currURI && prevURI !== currURI ? `<unmergable text regions of ${prevURI}, ${currURI}>` : prevURI ?? currURI;
+        result.fileURI = fileURI;
+    }
+    return result;
+}
```

Comment:
Nitpick: Missing newline at end of file.

Reply:
👍